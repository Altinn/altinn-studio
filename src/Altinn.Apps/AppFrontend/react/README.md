## Running the tests

### Lint checks

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run lint`.

### Unit tests

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run test`.
