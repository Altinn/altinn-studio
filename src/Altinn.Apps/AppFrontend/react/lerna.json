{
  "command": {
    "bootstrap": {
      "hoist": true
    }
  },
  "packages": [
    "./*"
  ],
  "version": "independent"
}
