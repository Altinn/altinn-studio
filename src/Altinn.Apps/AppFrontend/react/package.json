{
	"name": "app-frontend",
	"private": true,
	"scripts": {
		"install-deps": "lerna bootstrap --hoist --ci",
		"no-ci-hoist": "lerna bootstrap --hoist --no-ci",
		"install-deps-with-lock": "lerna bootstrap",
		"test": "lerna exec --concurrency 1 -- npm test",
		"install-all": "lerna exec -- npm ci",
		"clean-lock": "lerna exec -- rm ./package-lock.json",
		"clean-node-modules": "rm -rf ./node_modules && lerna exec --concurrency 1 -- rm -rf ./node_modules",
		"update-packages": "lerna exec --concurrency 1 -- npm update --dev",
		"list-outdated": "lerna run list-outdated"
	},
	"devDependencies": {
		"@typescript-eslint/eslint-plugin": "~4.8.2",
		"@typescript-eslint/parser": "~4.8.2",
		"eslint": "~7.14.0",
		"eslint-config-airbnb": "~18.2.1",
		"eslint-plugin-import": "~2.22.1",
		"eslint-plugin-jsx-a11y": "~6.4.1",
		"eslint-plugin-react": "~7.21.5",
		"eslint-plugin-react-hooks": "^4.2.0",
		"ini": "1.3.5",
		"lerna": "3.22.1"
	}
}
