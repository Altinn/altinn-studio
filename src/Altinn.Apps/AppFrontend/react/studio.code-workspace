{
	"folders": [
		{
			"path": "."
		}
  ],
  "settings": {
    "typescript.surveys.enabled": false,
    "typescript.tsdk": "node_modules/typescript/lib",
    "eslint.alwaysShowStatus": true,
    "eslint.validate": [
      "javascript",
      "javascriptreact",
      { "language": "typescript", "autoFix": true },
      { "language": "typescriptreact", "autoFix": true}
    ]
  }
}
