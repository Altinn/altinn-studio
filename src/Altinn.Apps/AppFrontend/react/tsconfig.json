{
  "compilerOptions": {
    "outDir": "compiled",
    "module": "commonjs",
    "target": "es5",
    "lib": [
      "es6",
      "dom",
      "es7",
      "esnext"
    ],
    "sourceMap": true,
    "allowJs": true,
    "jsx": "react",
    "moduleResolution": "node",
    "rootDir": "./",
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strictNullChecks": false,
    "noUnusedLocals": true
  },
  "include": [
    "**/*.ts",
    "**/*.tsx",
  ],
  "exclude": [
    "node_modules",
    "build",
    "scripts",
    "acceptance-tests",
    "webpack",
    "jest",
    "setupTests.ts",
    "templates",
    "**/setupTests.ts"
  ],
}
