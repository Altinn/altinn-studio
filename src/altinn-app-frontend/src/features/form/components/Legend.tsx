import React from 'react';
import type { ILabelSettings } from 'src/types';
import { LayoutStyle } from 'src/types';
import Description from './Description';
import { HelpTextContainer } from './HelpTextContainer';
import type { ILanguage } from 'altinn-shared/types';
import { RequiredIndicator } from './RequiredIndicator';
import { OptionalIndicator } from './OptionalIndicator';

export interface IFormLegendProps {
  labelText: React.ReactNode;
  descriptionText: React.ReactNode;
  language: ILanguage;
  required?: boolean;
  labelSettings?: ILabelSettings;
  helpText: React.ReactNode;
  id: string;
  layout?: LayoutStyle;
}

export default function Legend(props: IFormLegendProps) {
  if (!props.labelText) {
    return null;
  }
  const LabelText = (
    <>
      {props.labelText}
      <RequiredIndicator
        required={props.required}
        language={props.language}
      />
      <OptionalIndicator
        labelSettings={props.labelSettings}
        language={props.language}
        required={props.required}
      />
      {props.helpText && (
        <HelpTextContainer
          language={props.language}
          helpText={props.helpText}
        />
      )}
    </>
  );

  if (props.layout === LayoutStyle.Table) {
    return LabelText;
  }

  return (
    <>
      <label
        className='a-form-label title-label'
        htmlFor={props.id}
      >
        {LabelText}
      </label>
      {props.descriptionText && (
        <Description
          description={props.descriptionText}
          {...props}
        />
      )}
    </>
  );
}
