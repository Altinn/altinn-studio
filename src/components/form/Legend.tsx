import React from 'react';

import { PadlockLockedFillIcon } from '@navikt/aksel-icons';

import { Description } from 'src/components/form/Description';
import { HelpTextContainer } from 'src/components/form/HelpTextContainer';
import classes from 'src/components/form/Legend.module.css';
import { OptionalIndicator } from 'src/components/form/OptionalIndicator';
import { RequiredIndicator } from 'src/components/form/RequiredIndicator';
import { useLanguage } from 'src/features/language/useLanguage';
import { LayoutStyle } from 'src/layout/common.generated';
import type { ILabelSettings } from 'src/layout/common.generated';

export interface IFormLegendProps {
  label: React.ReactNode;
  description: React.ReactNode;
  required?: boolean;
  readOnly?: boolean;
  labelSettings?: ILabelSettings;
  helpText: React.ReactNode;
  id: string;
  layout?: LayoutStyle;
}

export function Legend({
  label,
  required,
  readOnly,
  labelSettings,
  id,
  helpText,
  description,
  layout,
}: IFormLegendProps) {
  const { elementAsString } = useLanguage();
  const labelAsText = elementAsString(label);
  if (!label || !labelAsText) {
    return null;
  }

  const LabelText = (
    <span className={classes.legendSpan}>
      {readOnly && (
        <PadlockLockedFillIcon
          aria-hidden={true}
          className={classes.padlock}
        />
      )}
      <span>
        {label}
        <RequiredIndicator required={required} />
        <OptionalIndicator
          labelSettings={labelSettings}
          required={required}
        />
      </span>
    </span>
  );

  if (layout === LayoutStyle.Table) {
    return LabelText;
  }

  return (
    <>
      <div className={classes.legendHelpTextContainer}>
        <legend>{LabelText}</legend>
        {helpText && (
          <HelpTextContainer
            helpText={helpText}
            title={labelAsText}
          />
        )}
      </div>
      {description && (
        <Description
          description={description}
          id={id}
        />
      )}
    </>
  );
}
