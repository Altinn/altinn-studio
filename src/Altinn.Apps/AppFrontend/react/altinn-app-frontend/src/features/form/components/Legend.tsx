/* eslint-disable react/prop-types */
import * as React from 'react';
import { ILabelSettings } from 'src/types';
import { getLanguageFromKey } from 'altinn-shared/utils';
import Description from './Description';
import { HelpTextContainer } from './HelpTextContainer';

export interface IFormLegendProps {
  labelText: string;
  descriptionText: string;
  language: any;
  required?: boolean;
  labelSettings?: ILabelSettings;
  helpText: string;
  id: string;
}

export default function Legend(props: IFormLegendProps) {
  if (!props.labelText) {
    return null;
  }

  return (
    <>
      <label
        className='a-form-label title-label'
        htmlFor={props.id}
      >
        {props.labelText}
        {(props.labelSettings?.optionalIndicator === false || props.required) ?
          null
          :
          <span className='label-optional'>
            ({getLanguageFromKey('general.optional', props.language)})
          </span>
        }
        {props.helpText &&
          <HelpTextContainer
            language={props.language}
            id={props.id}
            helpText={props.helpText}
          />
        }
      </label>
      {props.descriptionText &&
        <Description
          description={props.descriptionText}
          {...props}
        />
      }
    </>
  );
}
