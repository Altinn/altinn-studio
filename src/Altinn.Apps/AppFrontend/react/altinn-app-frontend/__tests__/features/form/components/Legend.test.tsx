import 'jest';
import * as React from 'react';
import '@testing-library/jest-dom/extend-expect';
import * as renderer from 'react-test-renderer';
import Legend from '../../../../src/features/form/components/Legend';

describe('>>> features/form/components/Legend.tsx', () => {
  beforeAll(() => {
  });

  beforeEach(() => {
  });

  it('+++ should render optional', () => {
    const rendered = renderer.create(
      <Legend
        id='legend1'
        labelText='legend.text'
        helpText=''
        language='no'
        required={false}
        descriptionText='legend.description'
      />,
    );
    expect(rendered).toMatchSnapshot();
  });

  it('+++ should not render optional because required', () => {
    const rendered = renderer.create(
      <Legend
        id='legend1'
        labelText='legend.text'
        helpText=''
        language='no'
        required={true}
        descriptionText='legend.description'
      />,
    );
    expect(rendered).toMatchSnapshot();
  });

  it('+++ should not render optional because labelSettings.optionalIndicator is false', () => {
    const rendered = renderer.create(
      <Legend
        id='legend1'
        labelText='legend.text'
        helpText=''
        language='no'
        required={false}
        labelSettings={{ optionalIndicator: false }}
        descriptionText='legend.description'
      />,
    );
    expect(rendered).toMatchSnapshot();
  });

  it('+++ should not render optional because required is true even when labelSettings.optionalIndicator is true', () => {
    const rendered = renderer.create(
      <Legend
        id='legend1'
        labelText='legend.text'
        helpText=''
        language='no'
        required={true}
        labelSettings={{ optionalIndicator: true }}
        descriptionText='legend.description'
      />,
    );
    expect(rendered).toMatchSnapshot();
  });
});
