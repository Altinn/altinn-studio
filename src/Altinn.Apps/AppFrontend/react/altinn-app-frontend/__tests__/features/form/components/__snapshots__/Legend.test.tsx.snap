// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> features/form/components/Legend.tsx +++ should not render optional because labelSettings.optionalIndicator is false 1`] = `
Array [
  <label
    className="a-form-label title-label"
    htmlFor="legend1"
  >
    legend.text
    
  </label>,
  <span
    className="a-form-label description-label"
    id="description-legend1"
  >
    legend.description
  </span>,
]
`;

exports[`>>> features/form/components/Legend.tsx +++ should not render optional because required 1`] = `
Array [
  <label
    className="a-form-label title-label"
    htmlFor="legend1"
  >
    legend.text
    
  </label>,
  <span
    className="a-form-label description-label"
    id="description-legend1"
  >
    legend.description
  </span>,
]
`;

exports[`>>> features/form/components/Legend.tsx +++ should not render optional because required is true even when labelSettings.optionalIndicator is true 1`] = `
Array [
  <label
    className="a-form-label title-label"
    htmlFor="legend1"
  >
    legend.text
    
  </label>,
  <span
    className="a-form-label description-label"
    id="description-legend1"
  >
    legend.description
  </span>,
]
`;

exports[`>>> features/form/components/Legend.tsx +++ should render optional 1`] = `
Array [
  <label
    className="a-form-label title-label"
    htmlFor="legend1"
  >
    legend.text
    <span
      className="label-optional"
    >
      (
      general.optional
      )
    </span>
    
  </label>,
  <span
    className="a-form-label description-label"
    id="description-legend1"
  >
    legend.description
  </span>,
]
`;
